### DISCUSSION TITLE
Why create a new Client on the fly for every request in getServerSideProps

### DISCUSSION THREAD

**Discussion by gszliu** [urql-graphql/urql#1521] [Q&A] [answered]
This is probably a very silly question. I'm very new to both the React and Next.js worlds. I'm building out an application with Next.js and querying a GraphQL endpoint for data, for now only on the server side (client-side in the future). I intend to make use of `getServerSideProps` to fetch data at the page level and compose or prop-drill as necessary to any children that need data.

In the [next-urql documentation](https://github.com/FormidableLabs/urql/tree/main/packages/next-urql#restricting-data-fetching-to-the-server), the suggestion for `getServerSideProps` is to use `initUrqlClient` to create a client on the fly - done once per request. In [the code](https://github.com/FormidableLabs/urql/blob/main/packages/next-urql/src/init-urql-client.ts#L13-L14) comments:

> Create a new Client for every server-side rendered request.
> This ensures we reset the state for each rendered page.

Why is this necessary? Is there no value in re-using a previously created Client? If every time a Client is created on the fly, is there any value in adding in exchanges like ssrExchange, or dedupExchange? For the latter, if I understand correctly, a newly created Client would not be aware of previous GraphQL requests made by the server, and not able to dedupe.

If the above is true, is it ideal to keep the server-side Client creation in `getServerSideProps` as simple as possible without adding exchanges, if they don't add value?

Edit: As a follow on question, there's also [this example](https://formidable.com/open-source/urql/docs/advanced/server-side-rendering/#using-getstaticprops-or-getserversideprops) of utilizing `ssrCache` and `ssrCache.extractData` with `getServerSideProps`. What is the advantage of this method over the other suggested method of creating on-the-fly Clients per request? When should I do this `ssrCache.extraData` methodology over the other?

Thanks in advance!

**Comment by JoviDeCroock** [comment, marked as answer]
The phrasing might be a bit off but we create a separate client just for the initial server-side render. This makes it so that we can do a bunch of requests on that client and it will only live in the closure of that initial data request. We should always utilize the `ssrCache.extractData()` method. This PR will shed more light on the different methods as next-urql has two ways of using it https://github.com/FormidableLabs/urql/pull/1502

**Reply by gszliu** [reply]
> We should always utilize the ssrCache.extractData() method. 

So if I’m understanding correctly, the simple approach suggested in the old next-urql example is not an ideal way to proceed because it doesn’t allow the ssrCache to fetch the data when hydrating on the client side. Dumb follow up - if I’m fully rendering a page on the server side, and it’s sent down the wire to a client’s browser, what is being hydrated? With full server side render, the page is not being re-rendered from scratch on the client side, right?

> we create a separate client just for the initial server-side render

As opposed to what exactly? Maybe I’m not understanding exactly how Next.js does SSR.

> can do a bunch of requests on that client and it will only live in the closure of that initial data request

So the purpose is in case we make multiple GraphQL requests for the same page. For example if we make the same identical request multiple times on one page for some reason, the dedupExchange would in fact kick in.

However, if page B makes an identical GraphQL request that page A made moments ago, because the Clients were created for each page’s gSSP, the dedupExchange added to Client for page B would not dedupExchange from page A’s request, correct?

If so, is there a way to dedupe across multiple pages? Or does that require a secondary / L2 cache.

**Reply by JoviDeCroock** [reply]
This is what `extractData` does, it looks at all requests made with the cilent you used on the server-side and makes sure they're hydrated on the client. You can double check this by checking the `dom-node` with `id="__NEXT_DATA__`.

Both examples are good, the automatic approach just uses `getInitialProps` and handles everything for you, as documented in the readme of #1502 if you want to use the `getXProps` functions you will need to do the manual work you alluded to in the original post.

**Reply by gszliu** [reply]
Could answer the other, dumber questions I had regarding rendering? If possible! Noob-friendlier, the better

**Reply by JoviDeCroock** [reply]
> This is what extractData does, it looks at all requests made with the cilent you used on the server-side and makes sure they're hydrated on the client. You can double check this by checking the dom-node with id="__NEXT_DATA__.

Basically if you fetch data on the server it's sent to that dom-node in question and the client-side cache will be hydrated with it, this means if you have a subsequent client-side transition that asks for that same data, it will not need to hit the server. Imagine it like a pipeline.

Our exchanges have `dedupe --> cache --> ssr --> fetch`, when we initially send this html to the client, the ssrExchange on the client will pick up the `__NEXT_DATA__` dom-node and hydrate its cache and send it to the actual cacheExchange when requests come in. When the ssrExchange doesn't have the data we need it'll send it to the fetchExchange to ask for it to the appropriate server.

**Reply by gszliu** [reply]
I see, so the data hydration is useful when adding client side navigation.

Lastly, if on the server side the Clients are created on the fly per request, do they properly take advantage of the cacheExchange? e.g. for the standard document cache which I assume is all in memory. Does a freshly created client with cacheExchange get access to another previous request’s cached data on the server side?

**Reply by JoviDeCroock** [reply]
A fresh client resets its cache but there can only be one `getServerSideProps` for a page so it being scoped to that one ssr-iteration is all you need. i.e. Caching cross-ssr-requests and having the cache be present would risk data-leaking across users.

**Reply by gszliu** [reply]
Thanks for the explanations! It makes more sense now.
